linera-core/src/updater.rs

```diff
@@ -221,25 +222,46 @@ where
         certificate: GenericCertificate<ConfirmedBlock>,
         delivery: CrossChainMessageDelivery,
     ) -> Result<Box<ChainInfo>, ChainClientError> {
-        let result = self
+        let mut result = self
             .remote_node
             .handle_optimized_confirmed_certificate(&certificate, delivery)
             .await;
 
-        Ok(match &result {
-            Err(original_err @ NodeError::BlobsNotFound(blob_ids)) => {
-                self.remote_node
-                    .check_blobs_not_found(&certificate, blob_ids)?;
-                // The certificate is confirmed, so the blobs must be in storage.
-                let maybe_blobs = self.local_node.read_blobs_from_storage(blob_ids).await?;
-                let blobs = maybe_blobs.ok_or_else(|| original_err.clone())?;
-                self.remote_node.node.upload_blobs(blobs.clone()).await?;
-                self.remote_node
-                    .handle_confirmed_certificate(certificate, delivery)
-                    .await
-            }
-            _ => result,
-        }?)
+        let mut sent_admin_chain = false;
+        let mut sent_blobs = false;
+        loop {
+            result = match result {
+                Err(NodeError::EventsNotFound(event_ids))
+                    if !sent_admin_chain
+                        && certificate.inner().chain_id() != self.admin_id
+                        && event_ids.iter().all(|event_id| {
+                            event_id.stream_id == StreamId::system(EPOCH_STREAM_NAME)
+                                && event_id.chain_id == self.admin_id
+                        }) =>
+                {
+                    // The validator doesn't have the committee that signed the certificate.
+                    sent_admin_chain = true;
+                    self.update_admin_chain().await?;
+                    self.remote_node
+                        .handle_confirmed_certificate(certificate.clone(), delivery)
+                        .await
+                }
+                Err(NodeError::BlobsNotFound(blob_ids)) if !sent_blobs => {
+                    // The validator is missing the blobs required by the certificate.
+                    sent_blobs = true;
```

Comment:
Maybe it doesn't matter in practice but I'd feel safer if we set it to `true` _after_ we actually upload the blobs. Similarly for the `sent_admin_chain` above.